### PR TITLE
v1.13

### PR DESCRIPTION
**Pull request by ppinck-CardTapp**
- Standardized naming convention of all classes and assets for consistency and to avoid potential naming conflicts
- Refactored CardtappAttachTapper and CardtappAttachTapperHelper to allow for handling of both Lead and Contact object types
- Added CardtappQueryBuilder class to support dynamic SOQL select statements
- Added notification preferences at the User level and support for notification flagging at Tapper creation
- Added workflow rules, email templates and email alerts for notification workflows

### CODE REVIEW

**Comment by ppinck-CardTapp** [comment]
Renaming the classes (and files) appears to have broken the association with the previous versions. I promise I didn't completely re-write every single class....

File: Salesforce_integration/classes/CardtappAttachTapper.cls [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    static final string EXISTING_LEAD_FOUND_WORKFLOW_ACTION_NAME = 'Existing Lead found';
+    static final string AMBIGUOUS_LEAD_CONTACT_FOUND_WORKFLOW_ACTION_NAME = 'Ambiguous Lead/Contact found';
+    
+    public static Cardtapp_Tapper__c AssociateUserToTapper(Cardtapp_Tapper__c t) {
+    	
+    	if (t.Member_external_id__c == null || t.Member_external_id__c == '') {
+    		return t;
+    	}
+    	
+    	string search_token = '%' + t.Member_external_id__c + '%';
```
Comment (RealNobody) [inline Salesforce_integration/classes/CardtappAttachTapper.cls]: Technically, I think that this string should be encoded not passed directly to the search.

I don't THINK that the external ID should ever have any values that will cause a problem, but I'm not a big cross my fingers and hope kind of guy.

**Review by RealNobody — APPROVED** [review]
👍

File: Salesforce_integration/classes/CardtappAttachTapperHelper.cls [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+        // See if a relevant ObjectId is provided with the Tapper record
+        if (isContact && !IsNullOrEmptyString(t.Salesforce_Contact__c)) {
+
+			filterMap = new Map<string,string> { 'Id' => t.Salesforce_Contact__c};			
+        	return_list = Database.query( CardtappQueryHelper.BuildLeadContactSearchQuery(object_type, filterMap,false));		
+			if (return_list.size() >= 1) { return return_list;}
+        }
+
+        if (isLead && !IsNullOrEmptyString(t.Salesforce_Lead__c)) {
```
Comment (dustinswisstack) [inline Salesforce_integration/classes/CardtappAttachTapperHelper.cls:24]: Indentation is not consistent. I would find an online formatter to get consistent:
http://stackoverflow.com/questions/206441/online-code-beautifier-and-formatter

File: Salesforce_integration/classes/CardtappAttachTapperHelperTest.cls [hunk 3]
```diff
@@ -58,31 +58,113 @@ private class AttachCardtappTapperHelperTest {
...
+		);
+		
+		CardtappAttachTapperHelper.UpdateSObjectFromTapperData(t1,c1);
+		CardtappAttachTapperHelper.UpdateSObjectFromTapperData(t2,c2);
+		CardtappAttachTapperHelper.UpdateSObjectFromTapperData(t3,c3);
+
+		Contact updated_c1 = [SELECT Id, Email, MobilePhone FROM Contact WHERE Id = :c1.Id];		
+		Contact updated_c2 = [SELECT Id, Email, MobilePhone FROM Contact WHERE Id = :c2.Id];		
+		Contact updated_c3 = [SELECT Id, Email, MobilePhone FROM Contact WHERE Id = :c3.Id];		
+
```
Comment (dustinswisstack) [inline Salesforce_integration/classes/CardtappAttachTapperHelperTest.cls:141]: you could make the above a method and pass :c1,:c2,:c3.Id as a parameter

**Review by dustinswisstack — APPROVED** [review]

**Pull request merged**